### PR TITLE
docs: add documentation for threading tuning parameters

### PR DESCRIPTION
**Pull request by sagarth**
## Pull Request Description
This PR adds documentation for tuning parameters related to multi-threading (MPI_THREAD_MULTIPLE).

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [ ] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
👍 Looks good. Ping me when it is ready to review.

File: doc/mpich/TUNING_PARAMETERS [hunk 1]
```diff
@@ -2,3 +2,32 @@ This document discusses some of the new features added to the MPICH-based MPI im
...
+Here are the standard hints from the MPI 4.0 standard that are supported (we recommend using both info hints if the application is not using wildcards):
+
+* `mpi_assert_no_any_source`: Set this info hint if the application does not use the wildcard MPI_ANY_SOURCE on the given communicator.
+* `mpi_assert_no_any_tag`: Set this hint if the application does not use MPI_ANY_TAG on the given communciator.
+
+MPICH also supports a non-standard per-communicator info hint that provides the user with more control of a communicator-VCI mapping:
+* `vci`: Specific sender and receiver VCI can be selected for each communicator using this hint. Applications can utilize this feature to control how messages are spread across the VCIs.
+For instance, programmers can write the applications so that pairs of threads communicate by using different communicator, where each communicator can be mapped to a different VCI through the `vci` info hint. Valid values for this info hints are 0 to (total vcis - 1). This hint is also utilized by MPICH to select a progress thread to process reduction collective operation.
```
Comment (hzhou) [inline doc/mpich/TUNING_PARAMETERS:33]: Could you wrap the text to width of less than 80 chars?

**Review by hzhou — CHANGES_REQUESTED** [review]

File: doc/mpich/TUNING_PARAMETERS (hunk 1 above)
Comment (sagarth) [reply doc/mpich/TUNING_PARAMETERS]: Done

File: doc/mpich/TUNING_PARAMETERS [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+contexts in the network card, which we call VCIs (Virtual Communication
+Interfaces).
+
+MPICH provides two threading synchronization strategies. One strategy is to use
+a `global` coarse-grained lock to protect every MPI call that accesses shared
+state inside the MPI runtime. With this model, if two threads call concurrently
+the MPI runtime, the runtime will serialize the calls. Another strategy uses a
+fine-grained lock to protect the access to each network context or VCI. With
+this approach, threads trying to access the same VCI get serialized, but threads
+accessing different VCIs can execute in parallel. With the fine-grained locking
```
Comment (hzhou) [inline doc/mpich/TUNING_PARAMETERS:40]: Need explain how to configure/select between "global" and "find-grained".

File: doc/mpich/TUNING_PARAMETERS [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+model, MPICH supports a `direct` mode where the MPI runtime itself grabs a per
+VCI lock and a `lockless` mode where MPI does not grab a lock and relies on the
+network provider to provide fine grain synchronization (MPICH uses the
+FI_THREAD_SAFE mode from OFI to support the `lockless` mode).
```
Comment (hzhou) [inline doc/mpich/TUNING_PARAMETERS:44]: Document the configure option and runtime cvars.

**Review by hzhou — CHANGES_REQUESTED** [review]
It's good explanations. However, it is immediately clear for a user how to configure/tune.  List configure options and cvars in separate code snippet block should help.

File: doc/mpich/TUNING_PARAMETERS (hunk 3 above)
Comment (sagarth) [reply doc/mpich/TUNING_PARAMETERS:44]: I'll add them.

File: doc/mpich/TUNING_PARAMETERS (hunk 2 above)
Comment (sagarth) [reply doc/mpich/TUNING_PARAMETERS:40]: I'll add the details.

Comment (sagarth) [reply doc/mpich/TUNING_PARAMETERS:40]: Done

File: doc/mpich/TUNING_PARAMETERS (hunk 3 above)
Comment (sagarth) [reply doc/mpich/TUNING_PARAMETERS:44]: Done

**Comment by sagarth** [comment]
@hzhou I've added a fixup patch with your requested changes. Could you check if it looks fine? I'll squash it after the review.

File: doc/mpich/TUNING_PARAMETERS [hunk 4]
```diff
@@ -43,12 +43,26 @@ VCI lock and a `lockless` mode where MPI does not grab a lock and relies on the
 network provider to provide fine grain synchronization (MPICH uses the
 FI_THREAD_SAFE mode from OFI to support the `lockless` mode).
 
+A threading synchronization strategy can be selected at the configuration time
+as:
+  --enable-thread-cs=global to select coarse-grained locking, and
+  --enable-thread-cs=per-vci to select fine-grained locking.
```
Comment (hzhou) [inline doc/mpich/TUNING_PARAMETERS:49]: Can you break the line so the command snippet is on its own line?

Comment (sagarth) [reply doc/mpich/TUNING_PARAMETERS:49]: I was debating with myself about it. Sure, let me do that.

Comment (sagarth) [reply doc/mpich/TUNING_PARAMETERS]: Done

File: doc/mpich/TUNING_PARAMETERS [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+information about per-communicator info hints that MPICH utilizes to make
+runtime decisions at a finer granularity.
+
+Notice that performance of the application depends on how ranks are mapped to
+the cores of the processor. For information on how to specify process-core
+binding with mpirun, information is provided here:
+https://wiki.mpich.org/mpich/index.php/Using_the_Hydra_Process_Manager#Process-core_Binding.
+
+
+
```
Comment (hzhou) [inline doc/mpich/TUNING_PARAMETERS:17]: Extra empty lines.

File: doc/mpich/TUNING_PARAMETERS [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+This following section of this document provides information about threading
+optimizations (Section 1).
+
+**1. Multi-threading**
+
+
```
Comment (hzhou) [inline doc/mpich/TUNING_PARAMETERS:23]: extra empty line

**Review by hzhou — APPROVED** [review]
I notice you are using MARKDOWN formating throughout. How about rename this document into `tuning_parameters.md` and then fix the part that doesn't look right (like headings)?

I am approving it as I am OK to merge it as is. It's up to you to make additional changes. Just let me know when you are happy for it to merge.

**Comment by sagarth** [comment]
Fixed the space issues. Also converted the doc file to `.md` and fixed some of the formatting issues. I kept the conversion on a separate patch though.

**Comment by hzhou** [comment]
How about add a title?

**Comment by sagarth** [comment]
> How about add a title?

Is the how we add title to a MARKDOWN file? : 
`% MPICH Tuning Parameters`

**Comment by hzhou** [comment]
> > How about add a title?
> 
> Is the how we add title to a MARKDOWN file? : `% MPICH Tuning Parameters`

Title is just bigger heading. 

# This is a title

## Section title

**Comment by sagarth** [comment]
Updated

**Pull request merged**